### PR TITLE
Improve button control by using primitive rectangles

### LINKED ISSUE #196: 🚧Improve button control by using primitive rectangles
### I have done the items below . . .

- [X] I have updated the title by replacing the '**_<title_**>' section.

### Description

Currently the button has to load a texture to display the face of the button.  This is expensive and not very customizable for a standard button.

Change the button to be able to use rectangles to render the face of the button. 

### Acceptance Criteria

**This issue is finished when:**
- [x] The button renders its face using rectangles instead of textures
- [x] Textures are removed from the project
- [x] The color of the face of the button is customizable
- [x] When the width or height of the button is changed, the label automatically centers itself
- [x] When the width or height of the button is smaller then the rectangle bounds of the label, the text does not render outside of the bounds of the button face.
- [x] The button has a `bool` border property that can be turned on and off
- [x] The button has a `Border Color` property
- [x] The button has a `Border Thickness` property
- [x] The button has a `CornerRadius` property
- [x] Add `AutoSize` property of type `bool` to the `Label` control
    - When true, the width and height of the `Label` will automatically be controlled based on the size of the font
    - If `AutoSize` is set to `false`, then the user can control the `Width` and `Height` of the control and only the characters contained within the bounds of will be rendered.
- [x] Add `AutoSize` property of type `bool` to the `Button` control
    - When true, the width and height of the `Button`  will automatically be controlled based on the size of the button `Label` text
    - If `AutoSize` is set to `false`, then the user can control the `Width` and `Height` of the control.
- [x] Unit tests added
- [x] All unit tests pass

### ToDo Items

- [x] Draft pull request created and linked to this issue
- [X] Priority label added to issue (**_low priority_**, **_medium priority_**, or **_high priority_**)
- [X] Issue linked to the proper project
- [X] Issue linked to proper milestone

### Issue Dependencies

_No response_

### Related Work

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>

<h2 style="font-weight:bold" align="center">✅Development Checklist✅</h2>

<details open><summary>🌳Branching🌳</summary>

Does the name of the head(source) branch for this pull request have the correct **_preview feature branch_** syntax?

Syntax: _preview/feature/&lt;issue-id&gt;-&lt;description&gt;_  
Example: _preview/feature/123-my-preview-change_
- [x] Yes
- [ ] No

Does the name of the base(target) branch for this pull request have the correct **_preview branch_** syntax?

Syntax: _preview/v&lt;major-number&gt;.&lt;minor-number&gt;.&lt;patch-number&gt;-preview.&lt;prev-number&gt;_  
Example: _preview/v1.2.3-preview.4_
- [x] Yes
- [ ] No

Is the head(source) branch for this pull request created from a branch with the correct **_preview branch_** syntax?

Syntax: _preview/v&lt;major-number&gt;.&lt;minor-number&gt;.&lt;patch-number&gt;-preview.&lt;prev-number&gt;_  
Example: _preview/v1.2.3-preview.4_
- [x] Yes
- [ ] No
</details>

<details open><summary>🐛Bugs🐛</summary>

Contains Bug Fix(es)
- [x] Yes
  - [x] A bug-label label has been added to the pull request.
- [ ] No
</details>

<details open><summary>💣Breaking Change(s)💣</summary>

Contains [breaking change(s)](https://docs.microsoft.com/en-us/dotnet/core/compatibility/#modifications-to-the-public-contract)
- [ ] Yes
  - [ ] A breaking-change-label label has been added to the pull request.
- [x] No
</details>

<details open><summary>✨Enhancements✨</summary>

Contains enhancement(s) that add a feature or behavior.
- [x] Yes
  - [x] An enhancement-label label has been added to the pull request.
- [ ]  No
</details>

<details open><summary>⚙️Workflow (CI/CD) Changes⚙️</summary>

Contains changes to workflow files. These changes can only be done by the project maintainers.
- [ ] Yes
  - [ ] A workflow-label label has been added to the pull request.
- [x] No
</details>

<details open><summary>📃Documentation Updates📃</summary>

Contains changes that require updates to existing **_code_** and/or **_Velaptor product_** documentation.
- [x] Yes
  - [x] I have updated the documentation accordingly.
    - Choose the type of documentation.
    - [x] Code Documentation
      - [x] A documentation-code-label label has been added to the pull request.
    - [ ] Product Documentation
      - [ ] A documentation-product-label label has been added to the pull request.
- [ ] No
</details>

<details open><summary>🧪Unit Testing🧪</summary>

My change requires unit tests to be written.
- [x] Yes
- [ ] No
</details>

<details open><summary>🧪Manual Testing🧪</summary>

I have manually tested my changes using the project named **_VelaptorTesting_**.
- [x] Yes
- [ ] No
</details>

<h2 style="font-weight:bold" align="center">✅Code Review Checklist✅</h2>

- [x] Pull request title matches the title of the linked issue.
- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.
- [x] My code follows the coding style of this project.
  - The style is enforced by the *.editorconfig* files in the project and displayed as warnings.  If there is an edge case with a coding style that should be ignored or changed, reach out, and let's discuss it.
- [x] A preview-label label has been added to the pull request.
- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vise-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - This does not mean that the tests pass in both the local and CI environment.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

**Review by CalvinWilkinson — COMMENTED** [review]
Requesting Changes

File: Testing/VelaptorTesting/Program.cs [hunk 1]
```diff
@@ -14,20 +15,54 @@ namespace VelaptorTesting
     /// </summary>
     public static class Program
     {
+        private const string ShowDebugConsole = "--show-debug-console";
+        private const string True = "true";
         private static IWindow? window;
         private static MainWindow? gameWindow;
 
-        public static void Main()
+        public static void Main(string[] args)
```
Comment (CalvinWilkinson) [inline Testing/VelaptorTesting/Program.cs:23]: Implement additional functionality.

- [ ] Implement a message that the application has started.
- [ ] Add some help text explaining the commands.
- [ ] Add another command to show help

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    // ReSharper restore RedundantNameQualifier
+
+    /// <summary>
+    /// A single item in a batch of glyph items that can be rendered to the screen.
+    /// </summary>
+    internal struct FontBatchItem : IEquatable<FontBatchItem>
+    {
+        /// <summary>
+        /// The source rectangle inside of the texture to render.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:23]: Refactor to:

```suggestion
        /// The source rectangle inside of the font atlas texture to render.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        public RectangleF SrcRect;
+
+        /// <summary>
+        /// The destination rectangular area of where to render the texture on the screen.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:28]: Refactor to:

```suggestion
        /// The destination rectangular area of where to render the font atlas texture on the screen.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    using System.Globalization;
+    using System.Text;
+    using Velaptor.Graphics;
+
+    // ReSharper restore RedundantNameQualifier
+
+    /// <summary>
+    /// A single item in a batch of glyph items that can be rendered to the screen.
+    /// </summary>
+    internal struct FontBatchItem : IEquatable<FontBatchItem>
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:20]: Refactor struct name to `FontGlyphBatchItem`

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        public RectangleF DestRect;
+
+        /// <summary>
+        /// The font glyph.
+        /// </summary>
+        public char Glyph;
+
+        /// <summary>
+        /// The size of the texture to be rendered.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:38]: Refactor to:

```suggestion
        /// The size of the glyph texture to be rendered.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        /// <remarks>This must be a value between 0 and 1.</remarks>
+        public float Size;
+
+        /// <summary>
+        /// The angle in degrees of the texture.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:44]: Refactor to:

```suggestion
        /// The angle in degrees of the glyph texture.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        /// <remarks>Needs to be a value between 0 and 360.</remarks>
+        public float Angle;
+
+        /// <summary>
+        /// The color to apply to the entire texture.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:50]: Refactor to:

```suggestion
        /// The color to apply to the entire glyph texture.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        public Color TintColor;
+
+        /// <summary>
+        /// The type of effects to apply to the texture when rendering.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:55]: Refactor to:

```suggestion
        /// The type of effects to apply to the glyph texture when rendering.
```

File: Velaptor/OpenGL/FontBatchItem.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        public RenderEffects Effects;
+
+        /// <summary>
+        /// The size of the viewport.
+        /// </summary>
+        public SizeF ViewPortSize;
+
+        /// <summary>
+        /// The ID of the texture.
```
Comment (CalvinWilkinson) [inline Velaptor/OpenGL/FontBatchItem.cs:65]: Refactor to:

```suggestion
        /// The ID of the font atlas texture.
```

File: Velaptor/Services/FontBatchService.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    // ReSharper disable RedundantNameQualifier
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using Velaptor.OpenGL;
+
+    // ReSharper restore RedundantNameQualifier
+
+    /// <summary>
+    /// Manages the process of batching up the rendering of text.
```
Comment (CalvinWilkinson) [inline Velaptor/Services/FontBatchService.cs:16]: Refactor to:

```suggestion
    /// Manages the process of batching up glyphs to be rendered..
```

File: Velaptor/Services/FontBatchService.cs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    /// </summary>
+    internal class FontBatchService : IBatchManagerService<FontBatchItem>
```
Comment (CalvinWilkinson) [inline Velaptor/Services/FontBatchService.cs:18]: Refactor name to `FontGlyphBatchService`

File: Velaptor/Services/FontBatchService.cs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        private SortedDictionary<uint, (bool shouldRender, FontBatchItem item)> batchItems = new ();
+        private uint currentBatchIndex;
+        private uint batchSize;
+        private uint currentFrame;
+
+        /// <summary>
+        /// Occurs when a batch is full.
+        /// </summary>
+        /// <remarks>
```
Comment (CalvinWilkinson) [inline Velaptor/Services/FontBatchService.cs:28]: Refactor remarks section to only include:

`The batch is ready when the total amount of items to be rendered is equal to the <see cref="BatchSize"/>.`

File: Velaptor/Services/FontBatchService.cs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            get => new (this.batchItems);
+            set => this.batchItems = new SortedDictionary<uint, (bool shouldRender, FontBatchItem item)>(value);
+        }
+
+        /// <summary>
+        /// Adds the given <paramref name="rect"/> to the batch.
+        /// </summary>
+        /// <param name="rect">The item to be added.</param>
+        public void Add(FontBatchItem rect)
```
Comment (CalvinWilkinson) [inline Velaptor/Services/FontBatchService.cs:63]: Refactor param name `rect` to `item`

**Review by CalvinWilkinson — COMMENTED** [review]
Changes requested

**Review by CalvinWilkinson — COMMENTED** [review]
PR Approved

**Review by kselena — CHANGES_REQUESTED** [review]

File: Testing/VelaptorTesting/Scenes/TestRenderTextScene.cs [hunk 14]
```diff
@@ -19,11 +19,14 @@ namespace VelaptorTesting.Scenes
     /// </summary>
     public class TestRenderTextScene : SceneBase
     {
+        private const int LeftMargin = 50;
+        private const int VertButtonSpacing = 10;
         private const string DefaultRegularFont = "TimesNewRoman-Regular.ttf";
         private const float AngularVelocity = 10f;
         private const float SizeChangeAmount = 0.5f;
         private const string SingleLineText = "Change me using the buttons to the left.";
         private const string MultiLineText = "Change me using\nthe buttons to the left.";
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply Testing/VelaptorTesting/Scenes/TestRenderTextScene.cs:28]: This is intended.  This is not a miss-spelling.

Anytime you see the value `\n` or `\t` in a string literal, this is an escape sequence to send the text after the escape sequence to next line, or tab before the space.

`\n` means new line
`\t` means a tab

File: Velaptor/Content/Fonts/IFont.cs [hunk 15]
```diff
@@ -96,5 +97,18 @@ public interface IFont : IContent
...
+        /// <summary>
+        /// Returns the bounds of each character in the given <paramref name="text"/> based on the
+        /// given <paramref name="textPos"/>.
+        /// </summary>
+        /// <param name="text">The text to get the bounds data for.</param>
+        /// <param name="textPos">The position of the text as a whole.</param>
+        /// <returns>The bounds for each character.</returns>
+        /// <remarks>
+        ///     The bounds includes the width, height, and position of the character relative to
+        ///     the <paramref name="textPos"/>.  The position is relative to the top left corner of the character.
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply Velaptor/Content/Fonts/IFont.cs:110]: This is correct and does not need to be changed.

The bounds contains a position and that position is relative to the top left corner of the glyph/character itself, not the center for example.

image

File: Velaptor/OpenGL/Buffers/RectGPUBuffer.cs [hunk 16]
```diff
@@ -288,5 +298,60 @@ private static RectGPUData ApplyColor(RectGPUData data, RectShape rect)
...
+        /// Process the border thickness by checking that the value is within limits.
+        /// If it is not within limits, it will force the value to be within limits.
+        /// </summary>
+        /// <param name="rect">The rectangle containing the border thickness to set within a limit.</param>
+        /// <remarks>
+        ///     This is done to prevent any undesired rendering artifacts from occuring.
+        /// </remarks>
+        private static RectShape ProcessBorderThicknessLimit(RectShape rect)
+        {
+            // TODO: This needs to be unit tested
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply Velaptor/OpenGL/Buffers/RectGPUBuffer.cs:312]: Yes.  Converted to an issue

File: Velaptor/OpenGL/Buffers/RectGPUBuffer.cs [hunk 17]
```diff
@@ -288,5 +298,60 @@ private static RectGPUData ApplyColor(RectGPUData data, RectShape rect)
...
+        /// are within limits.  If it is not within limits, it will force the values to be within limits.
+        /// </summary>
+        /// <param name="rect">The rectangle containing the radius values to process.</param>
+        /// <returns>The rect with the corner radius values set within limits.</returns>
+        /// <remarks>
+        ///     This is done to prevent any undesired rendering artifacts from occuring.
+        /// </remarks>
+        private static RectShape ProcessCornerRadiusLimits(RectShape rect)
+        {
+            // TODO: This needs to be unit tested
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply Velaptor/OpenGL/Buffers/RectGPUBuffer.cs:335]: Yes.  Converted to an issue

File: Velaptor/UI/TextBox.cs [hunk 18]
```diff
@@ -44,7 +44,7 @@ public TextBox(IContentLoader contentLoader)
 
             // TODO: Setup textbox border rect
```
[the remark this answers is not in the document]
Comment (CalvinWilkinson) [reply Velaptor/UI/TextBox.cs:45]: This entire control is subject to change and cannot be used by the user.

You can see this by looking at line 21.

``` cs
internal class TextBox : ControlBase
```

The keyword `internal` means that it is only accessible by the code in this library.  Any code outside of this library cannot use it and does not have access.  For example, the user using this library would not be able to use this entire class.

They would not be able to do the code below.

``` cs
var myTextBox = new TextBox();
```

**Pull request merged**